MacOS ci release mode build issue with gcc-14

The MacOS release-mode gcc-14 aarch64 CI started failing after this commit: https://github.com/oneapi-src/oneDNN/commit/55b52c67967b61d439c84df8b34cfb006bf528eb

The log can be seen [here](https://github.com/oneapi-src/oneDNN/actions/runs/11267787655/job/31333497703):

```
[ 31%] Building CXX object src/cpu/CMakeFiles/dnnl_cpu.dir/reorder/cpu_reorder_regular_f32_f16.cpp.o
In file included from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/common/dnnl_thread.hpp:24,
                 from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/simple_reorder.hpp:25,
                 from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/cpu_reorder.hpp:24,
                 from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/cpu_reorder_regular_f32_bf16.cpp:18:
In function 'void dnnl::impl::utils::array_copy(T*, const T*, size_t) [with T = long long int]',
    inlined from 'dnnl::impl::dim_t dnnl::impl::cpu::{anonymous}::get_quant_off(const dnnl_dim_t (&)[12], int, int, dnnl::impl::dim_t, dnnl::impl::dim_t, const dnnl::impl::memory_desc_t&)' at /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/simple_reorder.hpp:198:22,
    inlined from 'dnnl::impl::cpu::simple_reorder_impl<dnnl_f32, dnnl_format_tag_any, dnnl_bf16, dnnl_format_tag_any, true, dnnl::impl::cpu::spec::reference>::execute(const dnnl::impl::cpu::cpu_reorder_pd_t*, const dnnl::impl::exec_ctx_t&)::<lambda(dnnl::impl::dim_t)>' at /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/simple_reorder.hpp:2617:59:
/Users/runner/work/oneDNN/oneDNN/oneDNN/src/common/utils.hpp:215:16: error: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' specified bound between 18446744056529682432 and 18446744073709551608 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
  215 |         dst[i] = src[i];
      |         ~~~~~~~^~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [src/cpu/CMakeFiles/dnnl_cpu.dir/reorder/cpu_reorder_regular_f32_bf16.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/common/dnnl_thread.hpp:24,
                 from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/simple_reorder.hpp:25,
                 from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/cpu_reorder.hpp:24,
                 from /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/cpu_reorder_regular_f32_f16.cpp:17:
In function 'void dnnl::impl::utils::array_copy(T*, const T*, size_t) [with T = long long int]',
    inlined from 'dnnl::impl::dim_t dnnl::impl::cpu::{anonymous}::get_quant_off(const dnnl_dim_t (&)[12], int, int, dnnl::impl::dim_t, dnnl::impl::dim_t, const dnnl::impl::memory_desc_t&)' at /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/simple_reorder.hpp:198:22,
    inlined from 'dnnl::impl::cpu::simple_reorder_impl<dnnl_f32, dnnl_format_tag_any, dnnl_f16, dnnl_format_tag_any, true, dnnl::impl::cpu::spec::reference>::execute(const dnnl::impl::cpu::cpu_reorder_pd_t*, const dnnl::impl::exec_ctx_t&)::<lambda(dnnl::impl::dim_t)>' at /Users/runner/work/oneDNN/oneDNN/oneDNN/src/cpu/reorder/simple_reorder.hpp:2617:59:
/Users/runner/work/oneDNN/oneDNN/oneDNN/src/common/utils.hpp:215:16: error: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' specified bound between 18446744056529682432 and 18446744073709551608 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
  215 |         dst[i] = src[i];
      |         ~~~~~~~^~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [src/cpu/CMakeFiles/dnnl_cpu.dir/reorder/cpu_reorder_regular_f32_f16.cpp.o] Error 1
```

We suspect this is a compiler bug and attempts to reproduce it on [godbolt](https://godbolt.org/z/o7EnKG737) did not work, though I was able to reproduce on a local Mac machine. As a solution, we downgraded from gcc-14(.2.0) to gcc-13 in #2160. 

Leaving this here for future reference and debugging assistance, when the time comes to upgrade to gcc-14.

--- a/.github/workflows/ci-aarch64.yml
+++ b/.github/workflows/ci-aarch64.yml
@@ -24,25 +24,26 @@ on:
   pull_request:
     types: [opened, synchronize, reopened]
 
-# TODO: Fix job cancellation.
 #* Stop stale workflows when pull requests are updated: https://stackoverflow.com/a/70972844
 #* Does not apply to the main branch.
-# concurrency:
-#   group: ${{ github.ref }}
-#   cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}
 
 # Declare default permissions as read only.
 permissions: read-all
 
 jobs:
   macos:
-    name: macOS
     runs-on: macos-14
     strategy:
       matrix:
-        toolset: [clang, gcc]
-        config: [Debug, Release]
+        config: [
+          { toolset: clang, build: Release },
+          { toolset: gcc, build: Release }
+        ]
 
+    name: MacOS, ${{ matrix.config.toolset }}, ${{ matrix.config.build }}
     steps:
       - name: Checkout oneDNN
         uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
@@ -58,10 +59,10 @@ jobs:
         run: ${{ github.workspace }}/oneDNN/.github/automation/build_acl.sh
         env:
           ACL_ACTION: clone
-          ACL_CONFIG: ${{ matrix.config }}
+          ACL_CONFIG: ${{ matrix.config.build }}
           ACL_ROOT_DIR: ${{ github.workspace }}/ComputeLibrary
-          BUILD_TOOLSET: ${{ matrix.toolset }}
-          GCC_VERSION: 13
+          BUILD_TOOLSET: ${{ matrix.config.toolset }}
+          GCC_VERSION: 14
 
       - name: Get ACL commit hash for cache key
         id: get_acl_commit_hash
@@ -75,7 +76,7 @@ jobs:
         id: cache-acl-restore
         uses: actions/cache/restore@v4
         with:
-          key: ${{ steps.get_system_name.outputs.SystemName }}-acl-${{ matrix.toolset }}-${{ matrix.config }}-${{ steps.get_acl_commit_hash.outputs.ACLCommitHash }}
+          key: ${{ steps.get_system_name.outputs.SystemName }}-acl-${{ matrix.config.toolset }}-${{ matrix.config.build }}-${{ steps.get_acl_commit_hash.outputs.ACLCommitHash }}
           path: ${{ github.workspace }}/ComputeLibrary/build
 
       - name: Build ACL
@@ -84,9 +85,9 @@ jobs:
         env:
           ACL_ACTION: build
           ACL_ROOT_DIR: ${{ github.workspace }}/ComputeLibrary
-          BUILD_TOOLSET: ${{ matrix.toolset }}
-          ACL_CONFIG: ${{ matrix.config }}
-          GCC_VERSION: 13
+          BUILD_TOOLSET: ${{ matrix.config.toolset }}
+          ACL_CONFIG: ${{ matrix.config.build }}
+          GCC_VERSION: 14
 
       - name: Save ACL in cache
         if:  ${{ steps.cache-acl-restore.outputs.cache-hit != 'true' }}
@@ -101,16 +102,16 @@ jobs:
         working-directory: ${{ github.workspace }}/oneDNN
         env:
           ACL_ROOT_DIR: ${{ github.workspace }}/ComputeLibrary
-          BUILD_TOOLSET: ${{ matrix.toolset }}
-          CMAKE_BUILD_TYPE: ${{ matrix.config }}
-          GCC_VERSION: 13
+          BUILD_TOOLSET: ${{ matrix.config.toolset }}
+          CMAKE_BUILD_TYPE: ${{ matrix.config.build }}
+          GCC_VERSION: 14
     
-      - if: matrix.toolset == 'clang'
+      - if: matrix.config.toolset == 'clang'
         name: Run oneDNN smoke tests
         run: ${{ github.workspace }}/oneDNN/.github/automation/test_aarch64.sh
         working-directory: ${{ github.workspace }}/oneDNN/build
         env:
-          CMAKE_BUILD_TYPE: ${{ matrix.config }}
+          CMAKE_BUILD_TYPE: ${{ matrix.config.build }}
           DYLD_LIBRARY_PATH: ${{ github.workspace }}/ComputeLibrary/build
   
   # We only run the linux aarch64 runners if macos smoke tests pass.

